### PR TITLE
Detecting dynamic library injections

### LINKED ISSUE #10: Detect dynamic library injection
Detect dynamic library injection such as Shadow & Liberty Lite

### PR DESCRIPTION
**Pull request by hdsenevi**
Detecting dynamic library injections such as Liberty lite and shadow

### CODE REVIEW

File: JailMonkey/JailMonkey.m [hunk 1]
```diff
@@ -20,6 +36,323 @@ @implementation JailMonkey
...
+- (BOOL)isInjectedWithDynamicLibrary
+{
+    int i=0;
+    while(true){
+        const char *name = _dyld_get_image_name(i++);
+        if(name==NULL){
+            break;
+        }
+        if (name != NULL) {
+            char cyinjectHide[] = {
```
Comment (youran-studio) [inline JailMonkey/JailMonkey.m:91]: Seems we can pull all those constant out of the while loop

Comment (hdsenevi) [reply JailMonkey/JailMonkey.m:91]: you mean move these char arrays to a const file? yeah maybe that will keep this implementation file clean. Agreed, I'll do it.

File: JailMonkey/JailMonkey.m [hunk 2]
```diff
@@ -20,6 +36,323 @@ @implementation JailMonkey
...
+    NSLog(@"Found: %@", loc);
+}
+
+// Preventing libobjc hooked, strstr implementation
+const char* tuyul(const char* X, const char* Y)
+{
+    if (*Y == '\0')
+        return X;
+
+    for (int i = 0; i < strlen(X); i++)
```
Comment (youran-studio) [inline JailMonkey/JailMonkey.m:59]: this logic bit confuses for me, rarely see recurring function invoke with loop. 
Have you tested it, does is serve the purpose?

Comment (hdsenevi) [reply JailMonkey/JailMonkey.m:59]: It does. I've tested it. 

It's very elegant how it detects dynamic library injections when you think about it (and read the code). It is in a way strange code, but it does work like a charm. This function is the corner stone of the approach. 

Of cause I didn't come up with this. I lifted it from here : https://github.com/avltree9798/isJailbroken/blob/master/isJailbroken/JB.m#L74

File: JailMonkey/JailMonkey.m (hunk 1 above)
Comment (youran-studio) [reply JailMonkey/JailMonkey.m:91]: Oh, perhaps it's an overkill to create another constant file,  we could just lift all those constants out of the while loop.
on top of the while(true), maybe

File: JailMonkey/JailMonkey.m (hunk 2 above)
Comment (youran-studio) [reply JailMonkey/JailMonkey.m:59]: I think this function is supposed to match the substring  say,  X = "thisisasteak",  Y="steak".

but for this  X="AaBbCc" Y="abc"?    it is gonna return non-null for this as well.
I think We are looking for exact match, sorry I didn't test the func, may miss something for the code.

**Review by youran-studio — APPROVED** [review]
LGTM

**Pull request merged**